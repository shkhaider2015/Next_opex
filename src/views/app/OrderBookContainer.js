import React, { Component } from 'react';
import { Row } from 'reactstrap';
import openSocket from 'socket.io-client';
import { Colxx, Separator } from '../../components/common/CustomBootstrap';
import Breadcrumb from '../../containers/navs/Breadcrumb';
import OrderBook from '../../components/orderBook/OrderBook';
import { getHost } from '../../helpers/Utils';

export default class OrderBookContainer extends Component {
    constructor(props) {
        super(props);

        this.state = {
            askOrders: [],
            bidOrders: [],
        };
        this.socket = openSocket(getHost(''), {
            path: '/api/socket/data',
        });
        this.handleData = this.handleData.bind(this);
    }

    getBidAskOrders(data) {
        const order = [];

        Object.entries(data).forEach(
            ([key, value]) => {
                order.push({
                    price: Number(key),
                    quantity: Number(value),
                });
            },
        );

        return order;
    }

    handleOnLoadData(rawData) {
        const glass = JSON.parse(rawData[0].glass);

        this.setState({
            askOrders: this.getBidAskOrders(glass.asks),
            bidOrders: this.getBidAskOrders(glass.bids),
        });
    }

    handleData(rawData) {
        const obj = JSON.parse(rawData);
        const { glass } = obj[0];
        const glassObj = JSON.parse(glass);

        this.setState({
            askOrders: this.getBidAskOrders(glassObj.asks),
            bidOrders: this.getBidAskOrders(glassObj.bids),
        });
    }

    getOrderBookData() {
        try {
            const classInstance = this;
            const x = new XMLHttpRequest();
            x.open('GET', getHost('api/order-book'), true);
            x.onload = function() {
                const res = x.responseText && JSON.parse(x.responseText);
                classInstance.handleOnLoadData(res);
            };
            x.withCredentials = true;
            x.send();
        } catch (Error) {
            console.log('Error while fetching order book data : ', Error);
        }
    }

    componentDidMount() {
        this.getOrderBookData();
        this.socket.on('order-book:glass', this.handleData.bind(this));
    }

    componentWillUnmount() {
        this.socket.close();
    }

    render() {
        return (
            <>
                <Row>
                    <Colxx xxs="12">
                        <Breadcrumb heading="menu.order-book" match={this.props.match} />
                        <Separator className="mb-5" />
                  </Colxx>
              </Row>
                <Row>
                    <OrderBook askOrders={this.state.askOrders} bidOrders={this.state.bidOrders} />
              </Row>
          </>
        );
    }
}
