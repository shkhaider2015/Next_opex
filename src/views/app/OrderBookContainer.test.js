import React from 'react';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import OrderBookContainer from './OrderBookContainer';

configure({ adapter: new Adapter() });

describe('<OrderBookContainer />', () => {
    it('should render component with correct props', () => {
        const wrapper = shallow(<OrderBookContainer />);

        expect(wrapper.exists()).toBeTruthy();
    });

    it('should render component same as snapshot', () => {
        const wrapper = shallow(<OrderBookContainer />);

        expect(wrapper.getElements()).toMatchSnapshot();
    });
});
