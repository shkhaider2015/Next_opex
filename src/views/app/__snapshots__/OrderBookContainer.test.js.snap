// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<OrderBookContainer /> should render component same as snapshot 1`] = `
Array [
  <React.Fragment>
    <Row
      tag="div"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
      <Colxx
        xxs="12"
      >
        <BreadcrumbContainer
          heading="menu.order-book"
        />
        <Separator
          className="mb-5"
        />
      </Colxx>
    </Row>
    <Row
      tag="div"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
      <OrderBook
        askOrders={Array []}
        bidOrders={Array []}
      />
    </Row>
  </React.Fragment>,
]
`;

exports[`Start Should render chart 1`] = `
Array [
  <div
    className="row"
  >
    <div
      className="col-12"
    >
      <div
        className="separator mb-5"
      />
    </div>
  </div>,
  <div
    className="row"
  >
    <div
      className="mb-4 col-12"
    >
      <p>
        <span>
          Start
        </span>
      </p>
      <button
        aria-label={null}
        className="btn btn-secondary btn-lg"
        onClick={[Function]}
        type="button"
        variant="secondary"
      >
        5min
      </button>
      <button
        aria-label={null}
        className="btn btn-secondary btn-lg"
        onClick={[Function]}
        type="button"
        variant="secondary"
      >
        10min
      </button>
      <button
        aria-label={null}
        className="btn btn-secondary btn-lg"
        onClick={[Function]}
        type="button"
        variant="secondary"
      >
        Interactive
      </button>
    </div>
  </div>,
  <div
    className="row"
  >
    <div
      className="mb-34 col-12"
    />
  </div>,
  <div
    className="row"
  >
    <button
      aria-label={null}
      className="btn btn-secondary btn-lg"
      onClick={[Function]}
      type="button"
      variant="secondary"
    >
      Купить
    </button>
    <button
      aria-label={null}
      className="btn btn-secondary btn-lg"
      onClick={[Function]}
      type="button"
      variant="secondary"
    >
      Продать
    </button>
  </div>,
]
`;
